// (c) Copyright 2015 - 2023 Micro Focus or one of its affiliates..
/*
 * Created on 28/12/2015.
 */

var assert = require("assert");
var lftExpect = require("../../expect/lft_expect.js");
require('expectations');
var Q = require("q");

describe("lftExpect unit tests", function() {
    var promiseMock1;
    var promiseMock2;

    beforeEach(function() {
        promiseMock2 = {
            then: function(cb) {
                return cb(expect(7));
            }
        };

        promiseMock1 = {
            then: function() {
                return promiseMock2;
            }
        }
    });

    describe("construction", function () {
        it("should chain a promise with the expect if value passed is a promise", function() {
            var expectUnderTest = lftExpect(promiseMock1);
            assert.notEqual(expectUnderTest._promise, null);
        });

        it("should create regular expect if value passed is not a promise", function() {
            var expectUnderTest = lftExpect(8);
            assert.strictEqual(expectUnderTest.constructor.name, "Object");
        });

        it("should have all keys of expect except 'toThrow'", function() {
            var expectUnderTest = lftExpect(promiseMock1);
            var regularExpect = expect("");

            for(var expectKey in regularExpect) {
                if(expectKey !== "toThrow") {
                    assert(expectKey in expectUnderTest, "key: " + expectKey + " is not in expectUnderTest");
                }
            }
        });

        it("should not have the 'toThrow' key", function() {
            var expectUnderTest = lftExpect(promiseMock1);
            assert(!("toThrow" in expectUnderTest));
        });
    });

    describe("functions", function() {
        var deferred;
        var expectUnderTest;

        beforeEach(function() {
            deferred = Q.defer();
            expectUnderTest = lftExpect(deferred.promise);
        });

        describe("fail", function() {
            it("should throw exception with correct message", function(done) {
                deferred.resolve(null);
                var errMsg = "failing the test";
                expectUnderTest.fail(errMsg).then(
                    function() { done(errMsg);},
                    function(err) {assert.strictEqual(err.message, "expected null " + errMsg); done();}
                );
            });
        });

        //???
        //describe("pass", function() {
        //    it("should pass correct message", function() {
        //        deferred.resolve(null);
        //        return expectUnderTest.pass("test passed").then(function(msg) {
        //            assert.strictEqual(msg, "test passed");
        //        })
        //    });
        //});

        describe("toBe", function () {
            it("should succeed for correct value", function () {
                deferred.resolve(7);
                return expectUnderTest.toBe(7);
            });

            it("should fail for in correct value", function (done) {
                deferred.resolve(7);

                expectUnderTest.toBe(8).then(
                    function(){
                        done("exception was not thrown");},
                    function(){done();});
            });
        });

        describe("toBeCloseTo", function () {
            it("should succeed for correct value", function () {
                deferred.resolve(7.32);
                return expectUnderTest.toBeCloseTo(7.3,1);
            });

            it("should fail for in correct value", function (done) {
                deferred.resolve(7.32);

                expectUnderTest.toBeCloseTo(7.3,2).then(
                    function(){done("exception was not thrown");},
                    function(){done();});
            });
        });

        describe("toBeDefined", function () {
            it("should succeed for correct value", function () {
                deferred.resolve(7);
                return expectUnderTest.toBeDefined();
            });

            it("should fail for in correct value", function (done) {
                deferred.resolve(undefined);

                expectUnderTest.toBeDefined().then(
                    function(){done("exception was not thrown");},
                    function(){done();});
            });
        });

        describe("toBeFalsey", function () {
            it("should succeed for correct value", function () {
                deferred.resolve(null);
                return expectUnderTest.toBeFalsey();
            });

            it("should fail for in correct value", function (done) {
                deferred.resolve(true);

                expectUnderTest.toBeFalsey().then(
                    function(){done("exception was not thrown");},
                    function(){done();});
            });
        });

        describe("toBeFalsy", function () {
            it("should succeed for correct value", function () {
                deferred.resolve(null);
                return expectUnderTest.toBeFalsey();
            });

            it("should fail for in correct value", function (done) {
                deferred.resolve(true);

                expectUnderTest.toBeFalsey().then(
                    function(){done("exception was not thrown");},
                    function(){done();});
            });
        });

        describe("toBeGreaterThan", function () {
            it("should succeed for correct value", function () {
                deferred.resolve(7);
                return expectUnderTest.toBeGreaterThan(6);
            });

            it("should fail for in correct value", function (done) {
                deferred.resolve(7);

                expectUnderTest.toBeGreaterThan(8).then(
                    function(){done("exception was not thrown");},
                    function(){done();});
            });
        });

        describe("toBeLessThan", function () {
            it("should succeed for correct value", function () {
                deferred.resolve(7);
                return expectUnderTest.toBeLessThan(8);
            });

            it("should fail for in correct value", function (done) {
                deferred.resolve(7);

                expectUnderTest.toBeLessThan(6).then(
                    function(){done("exception was not thrown");},
                    function(){done();});
            });
        });

        describe("toBeNull", function () {
            it("should succeed for correct value", function () {
                deferred.resolve(null);
                return expectUnderTest.toBeNull();
            });

            it("should fail for in correct value", function (done) {
                deferred.resolve(7);

                expectUnderTest.toBeNull().then(
                    function(){done("exception was not thrown");},
                    function(){done();});
            });
        });

        describe("toBeTruthy", function () {
            it("should succeed for correct value", function () {
                deferred.resolve(true);
                return expectUnderTest.toBeTruthy();
            });

            it("should fail for in correct value", function (done) {
                deferred.resolve(null);

                expectUnderTest.toBeTruthy().then(
                    function(){done("exception was not thrown");},
                    function(){done();});
            });
        });

        describe("toBeUndefined", function () {
            it("should succeed for correct value", function () {
                deferred.resolve(undefined);
                return expectUnderTest.toBeUndefined();
            });

            it("should fail for in correct value", function (done) {
                deferred.resolve(null);

                expectUnderTest.toBeUndefined().then(
                    function(){done("exception was not thrown");},
                    function(){done();});
            });
        });

        describe("toContain", function () {
            it("should succeed for correct value", function () {
                deferred.resolve([2,3]);
                return expectUnderTest.toContain(2);
            });

            it("should fail for in correct value", function (done) {
                deferred.resolve([2,3]);

                expectUnderTest.toContain(4).then(
                    function(){done("exception was not thrown");},
                    function(){done();});
            });
        });

        describe("toEqual", function () {
            it("should succeed for correct value", function () {
                deferred.resolve({a: 1, b: 2});
                return expectUnderTest.toEqual({a: 1, b: 2});
            });

            it("should fail for in correct value", function (done) {
                deferred.resolve({a: 1, b: 2});

                expectUnderTest.toEqual({}).then(
                    function(){done("exception was not thrown");},
                    function(){done();});
            });
        });

        describe("toMatch", function () {
            it("should succeed for correct value", function () {
                deferred.resolve("how are you?");
                return expectUnderTest.toMatch(/you?/);
            });

            it("should fail for in correct value", function (done) {
                deferred.resolve("how are you?");

                expectUnderTest.toMatch(/aaa/).then(
                    function(){done("exception was not thrown");},
                    function(){done();});
            });
        });

        describe("toNotEqual", function () {
            it("should succeed for correct value", function () {
                deferred.resolve({a: 1, b: 2});
                return expectUnderTest.toNotEqual({});
            });

            it("should fail for in correct value", function (done) {
                deferred.resolve({a: 1, b: 2});

                expectUnderTest.toNotEqual({a: 1, b: 2}).then(
                    function(){done("exception was not thrown");},
                    function(){done();});
            });
        });
    });

    describe("report", function () {
        var deferred;
        var expectUnderTest;
        var reportMock = {};

        before(function() {
            // clean the cached lft_expect module
            global.__LFT_REPORTER__ = reportMock;
            var module = require.resolve("../../expect/lft_expect.js");
            var cachedModule = require.cache[module];
            delete require.cache[cachedModule.id];
            lftExpect = require("../../expect/lft_expect.js");
        });

        it("should report successful except event if value is promise", function(done) {
            reportMock._reportExpectEvent = function(funcName, hasNot, argumentsForReport, shouldThrow){
                assert.strictEqual(funcName, "toBe");
                assert.strictEqual(hasNot, null);
                assert.deepEqual(argumentsForReport, [7,7]);
                assert.strictEqual(shouldThrow, true);
            }

            deferred = Q.defer();
            expectUnderTest = lftExpect(deferred.promise);
            deferred.resolve(7);
            expectUnderTest.toBe(7).then(
                function(){done();},
                function(){done("exception was not thrown");});
        });

        it("should report successful not expect event if value is promise", function(done) {
            reportMock._reportExpectEvent = function(funcName, hasNot, argumentsForReport, shouldThrow){
                assert.strictEqual(funcName, "toBe");
                assert.strictEqual(hasNot, true);
                assert.deepEqual(argumentsForReport, [7,8]);
                assert.strictEqual(shouldThrow, true);
            }

            deferred = Q.defer();
            expectUnderTest = lftExpect(deferred.promise);
            deferred.resolve(7);
            expectUnderTest.not.toBe(8).then(
                function(){done();},
                function(){done("exception was thrown");});
        });

        it("should report failed expect event if value is promise", function(done) {
            reportMock._reportExpectEvent = function(funcName, hasNot, argumentsForReport, shouldThrow, e){
                assert.strictEqual(funcName, "toBe");
                assert.strictEqual(hasNot, null);
                assert.deepEqual(argumentsForReport, [8,7]);
                assert.strictEqual(shouldThrow, true);
                assert.notEqual(e, null);
            }

            deferred = Q.defer();
            expectUnderTest = lftExpect(deferred.promise);
            deferred.resolve(8);
            expectUnderTest.toBe(7).then(
                function(){done("exception was not thrown");},
                function(){done();});
        });

        it("should report successful expect event if value is not a promise", function() {
            reportMock._reportExpectEvent = function(funcName, hasNot, argumentsForReport, shouldThrow){
                assert.strictEqual(funcName, "toBe");
                assert.strictEqual(hasNot, null);
                assert.deepEqual(argumentsForReport, [7,7]);
                assert.strictEqual(shouldThrow, true);
            }

            expectUnderTest = lftExpect(7);
            expectUnderTest.toBe(7);
        });

        it("should report successful not expect event if value is not a promise", function() {
            reportMock._reportExpectEvent = function(funcName, hasNot, argumentsForReport, shouldThrow){
                assert.strictEqual(funcName, "toBe");
                assert.strictEqual(hasNot, true);
                assert.deepEqual(argumentsForReport, [7,8]);
                assert.strictEqual(shouldThrow, true);
            }

            expectUnderTest = lftExpect(7);
            expectUnderTest.not.toBe(8);
        });

        it("should report failed expect event if value is not a promise", function(done) {
            reportMock._reportExpectEvent = function(funcName, hasNot, argumentsForReport, shouldThrow, e){
                assert.strictEqual(funcName, "toBe");
                assert.strictEqual(hasNot, null);
                assert.deepEqual(argumentsForReport, [7,8]);
                assert.strictEqual(shouldThrow, true);
                assert.notEqual(e, null);
            }

            expectUnderTest = lftExpect(7);
            try {
                expectUnderTest.toBe(8);
            }catch(error){
                done();
                return;
            }
            done("exception wasn't thrown")
        });

        it("should report the event correct arguments from the same type", function() {
            var typesToCheck = [5,[0,1,2],{'key':'value'}];

            for(var i=0; i<typesToCheck.length; i++){
                reportMock._reportExpectEvent = function(funcName, hasNot, argumentsForReport){
                    assert.deepEqual(argumentsForReport, [typesToCheck[i],typesToCheck[i]]);
                }
                expectUnderTest = lftExpect(typesToCheck[i]);
                expectUnderTest.toBe(typesToCheck[i]);
            }
        });

        it("should report the event correct arguments from different type", function() {
            var expectedToCheck = [5,[0,1,2],{'key':'value'}];
            var actualToCheck = [[0,1,2],{'key':'value'},5];

            for(var i=0; i<expectedToCheck.length; i++){
                reportMock._reportExpectEvent = function(funcName, hasNot, argumentsForReport){
                    assert.deepEqual(argumentsForReport, [expectedToCheck[i],actualToCheck[i]]);
                }
                try {
                    expectUnderTest = lftExpect(expectedToCheck[i]);
                    expectUnderTest.toBe(actualToCheck[i]);
                }catch(error){
                    continue;
                }
            }
        });
    });
});